### PR TITLE
OHRM5X-976: Project Activity unique name validation API added

### PR DESCRIPTION
**Pull request by VageeshOrangeHRM**
This PR Contains following changes:
1. Unique validation added for project activity name.
2. Unit test added for the API.
3. Data group permission added for the API.
4. Unit test covered for customer DAO layer.

### CODE REVIEW

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 1]
```diff
@@ -143,4 +143,25 @@ public function deleteProjectActivities(array $toBeDeletedActivityIds): int
...
     }
+
+    /**
+     **this function for validating the project activity name availability. ( true -> project activity name already exist, false - project activity name is not exist )
+     * @param int $projectId
+     * @param string $projectActivityName
+     * @param int|null $projectActivityId
+     * @return bool
+     */
+    public function isProjectNameTaken(int $projectId, string $projectActivityName, ?int $projectActivityId = null): bool
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php:154]:
```suggestion
    public function isProjectActivityNameTaken(int $projectId, string $projectActivityName, ?int $projectActivityId = null): bool
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php [hunk 2]
```diff
@@ -143,4 +143,25 @@ public function deleteProjectActivities(array $toBeDeletedActivityIds): int
...
+        $q = $this->createQueryBuilder(ProjectActivity::class, 'projectActivity');
+        $q->andWhere('projectActivity.name = :projectActivityName');
+        $q->andWhere('projectActivity.project = :projectId');
+        $q->setParameter('projectActivityName', $projectActivityName);
+        $q->setParameter('projectId', $projectId);
+        if (!is_null($projectActivityId)) {
+            $q->andWhere('projectActivity.id != :projectActivityId'); // we need to skip the current project activity Name on update, otherwise count always return 1
+            $q->setParameter('projectActivityId', $projectActivityId);
+        }
+        return $this->getPaginator($q)->count() === 0;
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php]: done

File: symfony/plugins/orangehrmTimePlugin/Api/ValidationProjectActivityNameAPI.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        );
+
+        if (!is_null($projectActivityId)) {
+            $projectActivity = $this->getProjectActivityService()
+                ->getProjectActivityDao()
+                ->getProjectActivityByProjectIdAndProjectActivityId($projectId, $projectActivityId);
+            $this->throwRecordNotFoundExceptionIfNotExist($projectActivity, ProjectActivity::class);
+        }
+
+        $isChangeableProjectActivityName = $this->getProjectActivityService()
```
[the remark this answers is not in the document]
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/ValidationProjectActivityNameAPI.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php (hunk 1 above)
Comment (VageeshOrangeHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/ProjectActivityDao.php]: done

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/UserDaoTest.php [hunk 4]
```diff
@@ -206,7 +206,7 @@ public function testGetEmployeesByUserRole(): void
 
     public function testUserStatusByUserName(): void
     {
-        $status = $this->systemUserDao->isUserNameExistByUserName('samantha');
+        $status = !$this->systemUserDao->isUserNameExistByUserName('samantha');
         $this->assertEquals(false, $status);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/UserDaoTest.php:210]:
```suggestion
        $status = $this->systemUserDao->isUserNameExistByUserName('samantha');
        $this->assertTrue($status);
```

File: symfony/plugins/orangehrmPimPlugin/test/Dao/EmployeeDaoTest.php [hunk 5]
```diff
@@ -339,13 +339,13 @@ public function testGetEmailList(): void
...
-        $status = $this->employeeDao->isWorkEmailAvailableByCurrentEmail('kayla0001@xample.com', 'kayla@xample.com');
+        $status = !$this->employeeDao->isWorkEmailAvailableByCurrentEmail('kayla0001@xample.com', 'kayla@xample.com');
         $this->assertEquals(true, $status);
     }
 
     public function testIsWorkEmailAvailable(): void
     {
-        $status = $this->employeeDao->isWorkEmailAvailable('renukshan@xample.com');
+        $status = !$this->employeeDao->isWorkEmailAvailable('renukshan@xample.com');
         $this->assertEquals(false, $status);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/test/Dao/EmployeeDaoTest.php:349]:
```suggestion
        $status = $this->employeeDao->isWorkEmailAvailable('renukshan@xample.com');
        $this->assertTrue($status);
```

File: symfony/plugins/orangehrmTimePlugin/test/Api/ValidationProjectActivityNameAPITest.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * @group APIv2
+ */
+class ValidationProjectActivityNameAPITest extends EndpointIntegrationTestCase
+{
+    /**
+     * @dataProvider dataProviderForTestGetOne
+     */
+    public function testGetOne(TestCaseParams $testCaseParams): void
+    {
+        $this->populateFixtures('ProjectActivityAPI.yaml');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/test/Api/ValidationProjectActivityNameAPITest.php:38]: Preferred if you can use the same name of the class. e.g. `ValidationProjectActivityNameAPITest.yaml`

File: symfony/plugins/orangehrmTimePlugin/test/Api/ValidationProjectActivityNameAPITest.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        $this->createKernelWithMockServices([Services::AUTH_USER => $this->getMockAuthUser($testCaseParams)]);
+        $this->registerServices($testCaseParams);
+        $this->registerMockDateTimeHelper($testCaseParams);
+        $api = $this->getApiEndpointMock(ValidationProjectActivityNameAPI::class, $testCaseParams);
+        $this->assertValidTestCase($api, 'getOne', $testCaseParams);
+    }
+
+    public function dataProviderForTestGetOne(): array
+    {
+        return $this->getTestCases('ProjectActivityNameTestCase.yaml', 'GetOne');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/test/Api/ValidationProjectActivityNameAPITest.php:48]: same here. e.g. `ValidationProjectActivityNameAPITestCases.yaml`

File: symfony/plugins/orangehrmTimePlugin/test/Dao/CustomerDaoTest.php [hunk 8]
```diff
@@ -86,6 +96,20 @@ public function testGetCustomerById(): void
...
+        $result = $this->customerDao->getCustomerById(10);
+        $this->assertFalse($result instanceof Customer);
+        $this->assertEquals(null, $result);
+    }
+
+    public function testGetCustomer(): void
+    {
+        $result = $this->customerDao->getCustomer(100);
+        $this->assertFalse($result instanceof Customer);
+        $this->assertEquals(null, $result);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/test/Dao/CustomerDaoTest.php:110]:
```suggestion
        $this->assertNull($result);
```

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**